### PR TITLE
Change only_local to a enum type.

### PR DESCRIPTION
**Pull request by surechen**
Change only_local to enum type and change the macros to always require a variant of that enum.

r? @lcnr

### CODE REVIEW

**Comment by surechen** [comment]
> you need to rebase to the newest master branch

Done. Thank you.

**Comment by Nilstrieb** [comment]
Can we invert it and call it `EncodeCrossCrate` instead? I think that makes it more directly clear what it means.

**Comment by surechen** [comment]
> Can we invert it and call it `EncodeCrossCrate` instead? I think that makes it more directly clear what it means.

Agreed, `EncodeCrossCrate` is clearer to me, I'll adjust it. Thank you.

File: compiler/rustc_feature/src/builtin_attrs.rs [hunk 1]
```diff
@@ -253,12 +243,18 @@ macro_rules! experimental {
...
+pub enum EncodeCrossCrate {
+    Yes,
+    No,
+}
+
 pub struct BuiltinAttribute {
     pub name: Symbol,
     /// Whether this attribute is only used in the local crate.
     ///
     /// If so, it is not encoded in the crate metadata.
```
Comment (lcnr) [inline compiler/rustc_feature/src/builtin_attrs.rs:256]: please also update the polarity of this commit :grin:

File: compiler/rustc_feature/src/builtin_attrs.rs [hunk 2]
```diff
@@ -1004,10 +1069,13 @@ pub fn is_builtin_attr_name(name: Symbol) -> bool {
     BUILTIN_ATTRIBUTE_MAP.get(&name).is_some()
 }
 
-/// Whether this builtin attribute is only used in the local crate.
+/// Whether this builtin attribute is encoded cross crate.
+/// `EncodeCrossCrate::No` means it is only used in the local crate.
 /// If so, it is not encoded in the crate metadata.
```
Comment (lcnr) [inline compiler/rustc_feature/src/builtin_attrs.rs:1074]:
```suggestion
```

File: compiler/rustc_middle/src/ty/mod.rs [hunk 3]
```diff
@@ -1752,8 +1752,11 @@ impl<'tcx> TyCtxt<'tcx> {
...
         if let Some(did) = did.as_local() {
             self.hir().attrs(self.local_def_id_to_hir_id(did)).iter().filter(filter_fn)
-        } else if cfg!(debug_assertions) && rustc_feature::is_builtin_only_local(attr) {
-            bug!("tried to access the `only_local` attribute `{}` from an extern crate", attr);
+        } else if cfg!(debug_assertions) && !rustc_feature::is_builtin_encode_cross_crate(attr) {
+            bug!(
+                "tried to access the `not encode cross crate` attribute `{}` from an extern crate",
+                attr
+            );
         } else {
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/mod.rs:1760]:
```suggestion
        } else {
            debug_assert!(rustc_feature::is_builtin_encode_cross_crate(attr));
```

**Review by lcnr — COMMENTED** [review]
r=me after nits, thanks!

File: compiler/rustc_feature/src/builtin_attrs.rs [hunk 4]
```diff
@@ -1004,10 +1069,13 @@ pub fn is_builtin_attr_name(name: Symbol) -> bool {
...
-pub fn is_builtin_only_local(name: Symbol) -> bool {
-    BUILTIN_ATTRIBUTE_MAP.get(&name).is_some_and(|attr| attr.only_local)
+pub fn is_builtin_encode_cross_crate(name: Symbol) -> bool {
+    BUILTIN_ATTRIBUTE_MAP.get(&name).is_some_and(|attr| {
```
Comment (lcnr) [inline compiler/rustc_feature/src/builtin_attrs.rs:1076]: you flip the condition but still use `is_some_and` :thinking: this means that attributes which are *not* builtin now never get encoded in the crate metadata.

File: compiler/rustc_middle/src/ty/mod.rs [hunk 5]
```diff
@@ -1787,8 +1790,14 @@ impl<'tcx> TyCtxt<'tcx> {
...
-        if cfg!(debug_assertions) && !did.is_local() && rustc_feature::is_builtin_only_local(attr) {
-            bug!("tried to access the `only_local` attribute `{}` from an extern crate", attr);
+        if cfg!(debug_assertions)
+            && !did.is_local()
+            && !rustc_feature::is_builtin_encode_cross_crate(attr)
+        {
+            bug!(
+                "tried to access the `not encode cross crate` attribute `{}` from an extern crate",
+                attr
+            );
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/mod.rs:1800]: this check is redundant :thinking: we check this in `get_attrs` already.

while you're touching this code, please remove it :grin:

File: compiler/rustc_feature/src/builtin_attrs.rs (hunk 4 above)
Comment (lcnr) [reply compiler/rustc_feature/src/builtin_attrs.rs:1077]: :sweat_smile: I dislike this function name. It's ambiguous whether it's `!is(builtin) && is(encode_cross_crate)`  or `!is(builtin) || is(encode_cross_crate)` or even `!(is(builtin) && is(encode_cross_crate))`.

I also think we should generally avoid having negation in the name of a function returning a `bool`.

keep it as `fn is_encode_cross_crate` and change it's body to
```rust
pub fn is_encode_cross_crate(name: Symbol) -> bool {
    if let Some(attr) = BUILTIN_ATTRIBUTE_MAP.get(&name) {
        if attr.encode_cross_crate == EncodeCrossCrate::Yes { false } else { true }
    } else {
        true
    }
}
```

Comment (surechen) [reply compiler/rustc_feature/src/builtin_attrs.rs]: > 😅 I dislike this function name. It's ambiguous whether it's `!is(builtin) && is(encode_cross_crate)` or `!is(builtin) || is(encode_cross_crate)` or even `!(is(builtin) && is(encode_cross_crate))`.
> 
> I also think we should generally avoid having negation in the name of a function returning a `bool`.
> 
> keep it as `fn is_encode_cross_crate` or even just `fn encode_cross_crate` and change it's body to
> 
> ```rust
> pub fn encode_cross_crate(name: Symbol) -> bool {
>     if let Some(attr) = BUILTIN_ATTRIBUTE_MAP.get(&name) {
>         if attr.encode_cross_crate == EncodeCrossCrate::Yes { false } else { true }
>     } else {
>         true
>     }
> }
> ```

Done. Thank you.

**Pull request merged**